### PR TITLE
Add `numblocks=` argument to `normalize_chunks`

### PR DESCRIPTION
**Pull request by gjoseph92**
Precursor to https://github.com/dask/dask/issues/9412. This seemed more useful than inlining all the necessary logic into `from_delayed`.

cc @ian-r-rose @jrbourbeau @jsignell

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -3010,37 +3014,80 @@ def normalize_chunks(chunks, shape=None, limit=None, dtype=None, previous_chunks
...
+        else:
+            raise ValueError(CHUNKS_NONE_ERROR_MESSAGE)
     if isinstance(chunks, list):
         chunks = tuple(chunks)
     if isinstance(chunks, (Number, str)):
-        chunks = (chunks,) * len(shape)
+        if ndim is None:
+            raise ValueError(
+                "Must give `shape` or `numblocks` if chunks is not a tuple of tuples."
+            )
```
Comment (ian-r-rose) [inline dask/array/core.py:3052]: Nice additional coverage

**Review by ian-r-rose — COMMENTED** [review]
I must say, making `normalize_chunks` richer fills me with a lot of trepidation -- I already think it's an extremely difficult function to read and reason about, and I would not be at all surprised to find untested-but-valid combinations of the arguments that don't work.

I'm coming to this from the experience of spending a couple of evenings trying to add static type analysis here (and came up with some of the same additional `ValueErrors` that you found). I never wound up opening a PR, though:

https://github.com/dask/dask/compare/main...ian-r-rose:what-are-chunks?expand=1

**Comment by gjoseph92** [comment]
I agree with your trepidation. I tried to avoid doing this, but my implementation without it ended up copying so much from `normalize_chunks` that I grudgingly made this change. You can take a look at https://github.com/dask/dask/pull/9417 and see if you can think of a way to avoid it, but supporting the many possibilities of numeric chunk, tuple chunksize, explicit tuple of tuples, mixed tuple of ints and tuple of tuples, all of that with the possibility of lists or ndarrays, was not something I wanted to write test coverage for :)

I was hoping that this might at least be useful to something somewhere else in the future.

File: dask/array/core.py [hunk 2]
```diff
@@ -3010,37 +3014,80 @@ def normalize_chunks(chunks, shape=None, limit=None, dtype=None, previous_chunks
...
+    ndim: int | None = None
+    if shape is not None:
+        ndim = len(shape)
+    if numblocks is not None:
+        if ndim is not None:
+            if len(numblocks) != ndim:
+                raise ValueError(
+                    f"`shape` indicates {ndim} dimension(s), but `numblocks` indicates {len(numblocks)}."
+                )
+        ndim = len(numblocks)
```
Comment (jsignell) [inline dask/array/core.py:3038]: I'm wondering if this whole bit could be separated out into a separate function maybe called `_get_ndim` or something.

File: dask/array/core.py [hunk 3]
```diff
@@ -3010,37 +3014,80 @@ def normalize_chunks(chunks, shape=None, limit=None, dtype=None, previous_chunks
...
+    Note that ``'auto'``, -1, and None cannot be used if only ``numblocks`` is given.
+
+    >>> normalize_chunks(4, numblocks=(2, 3))
+    ((4, 4), (4, 4, 4))
+
+    Use ``shape`` and ``numblocks`` without any chunksize
+
+    >>> normalize_chunks(None, shape=(10, 5), numblocks=(2, 2))
+    ((5, 5), (3, 2))
     """
```
Comment (jsignell) [inline dask/array/core.py:3028]: It might be helpful to move some error handling to the very top. Like if all of `numblocks`, `shape`, and `chunks` are None, that is invalid. Or if `numblocks` and `shape` are None and chunks is not an iterable. Basically I think if you can separate out the error handling from the transformation that would be easier to parse.

**Review by jsignell — COMMENTED** [review]
I'm wondering if this could be split into several sub-functions like:

`_with_shape_and_chunks`, `_with_numblocks_and_chunks`, `with_shape_and_numblocks`, `_with_shape_and_numblocks_and_chunks`, `_with_just_chunks`.

I think what makes this function hard to parse is that it is so permissive that at every point you need to check that basically every field exists.
